### PR TITLE
Tweak: AI - Open web app when prompt is URL [ED-13130]

### PR DESCRIPTION
**Pull request by DennisNerush**
Tweak: Open web app when prompt is url [ED-13130]

[ED-13130]: https://elementor.atlassian.net/browse/ED-13130?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/ai/assets/js/editor/pages/form-layout/index.js [hunk 1]
```diff
@@ -121,6 +125,10 @@ const FormLayout = ( {
 			return;
 		}
 
+		if ( isURL( prompt ) ) {
+			setShouldRenderWebApp( true );
```
Comment (matipojo) [inline modules/ai/assets/js/editor/pages/form-layout/index.js:129]:
```suggestion
			setShouldRenderWebApp( true );
			return;
```

**Review by matipojo — CHANGES_REQUESTED** [review]
Please cover with tests

File: modules/ai/assets/js/editor/pages/form-layout/components/attachments/url-dialog.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ import { AlertDialog } from '../../../../components/alert-dialog';
 import { useTimeout } from '../../../../hooks/use-timeout';
 
 export const UrlDialog = ( props ) => {
-	const { iframeSource } = useAttachUrlService();
+	const { iframeSource } = useAttachUrlService( { url: props.url } );
```
Comment (matipojo) [inline modules/ai/assets/js/editor/pages/form-layout/components/attachments/url-dialog.js:10]: Maybe: 

```suggestion
	const { iframeSource } = useAttachUrlService( { targetUrl: props.url } );
```

File: modules/ai/assets/js/editor/pages/form-layout/index.js [hunk 3]
```diff
@@ -254,7 +262,15 @@ const FormLayout = ( {
 							onCancel={ () => setShowUnsavedChangesAlert( false ) }
 						/>
 					) }
-
+					{ shouldRenderWebApp && (
+						<AttachDialog
+							type={ ATTACHMENT_TYPE_URL }
+							url={ promptInputRef.current.value }
+							onAttach={ () => {} }
```
Comment (matipojo) [inline modules/ai/assets/js/editor/pages/form-layout/index.js:269]: It should preview the attachment, etc.

```suggestion
							onAttach={ onAttach }
```

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	beforeEach( async () => {
+		global.elementorCommon = elementorCommon;
+
+		global.ResizeObserver =
+			global.ResizeObserver ||
+			jest.fn().mockImplementation( () => ( {
+				disconnect: jest.fn(),
+				observe: jest.fn(),
+				unobserve: jest.fn(),
+			} ) );
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:23]: Extract to a reusable mock function

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	} );
+
+	it( 'Should render AttachDialog iframe once prompt is url', async () => {
+		const { getByTestId } = renderElement();
+
+		await addPromptAndGenerate( 'https://www.google.com' );
+		await sleep( 1000 );
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:30]: Try to remove

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+		expect( root.querySelector( 'iframe' ) ).not.toBeNull();
+		expect( root.querySelector( 'iframe' ).src ).toContain( 'https%3A%2F%2Fwww.google.com' );
+	} );
+
+	it( 'Should not render AttachDialog iframe when prompt is not url', async () => {
+		const { getByTestId } = renderElement();
+
+		await addPromptAndGenerate( 'How are you doing?' );
+		await sleep( 1000 );
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:42]: Try to remove

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js (hunk 5 above)
Comment (DennisNerush) [reply tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:30]: doesn't work without the sleep.

File: tests/jest/unit/modules/ai/assets/js/editor/layout-app.test.js [hunk 7]
```diff
@@ -42,14 +49,7 @@ describe( 'LayoutApp', () => {
 	let rerender;
 	beforeEach( async () => {
 		global.elementorCommon = elementorCommon;
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/layout-app.test.js:51]: Maybe move to `mockEditorEnvironment`
```suggestion
```

File: tests/jest/unit/modules/ai/assets/js/editor/layout-app.test.js [hunk 8]
```diff
@@ -42,14 +49,7 @@ describe( 'LayoutApp', () => {
...
-
-		global.ResizeObserver =
-			global.ResizeObserver ||
-			jest.fn().mockImplementation( () => ( {
-				disconnect: jest.fn(),
-				observe: jest.fn(),
-				unobserve: jest.fn(),
-			} ) );
+		mockResizeObserver();
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/layout-app.test.js:52]: Maybe
```suggestion
		mockEditorEnvironment();
```

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import TestThemeProvider from '../../mock/test-theme-provider';
+import { addPromptAndGenerate, mockResizeObserver, sleep } from '../../test-utils';
+import {
+	RemoteConfigProvider,
+} from 'elementor/modules/ai/assets/js/editor/pages/form-layout/context/remote-config';
+
+describe( 'FormLayout', () => {
+	beforeEach( async () => {
+		global.elementorCommon = elementorCommon;
+		mockResizeObserver();
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:16]:
```suggestion
		mockEditorEnvironment();
```

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+		expect( root.querySelector( 'iframe' ) ).not.toBeNull();
+		expect( root.querySelector( 'iframe' ).src ).toContain( 'https%3A%2F%2Fwww.google.com' );
+	} );
+
+	it( 'Should not render AttachDialog iframe when prompt is not url', async () => {
+		const { getByTestId } = renderElement();
+
+		await addPromptAndGenerate( 'How are you doing?' );
+		// Await sleep( 1000 );
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:35]:
```suggestion
```

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+						onClose={ () => {} }
+						onConnect={ () => {} }
+						onData={ () => {} }
+						onInsert={ () => {} }
+						onSelect={ () => {} }
+						onGenerate={ () => {} }>
+						<FormLayout { ...props } />
+					</ConfigProvider>
+				</RemoteConfigProvider>
+			</TestThemeProvider>,
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:63]: Maybe extract to a `TestWrapper` including the `mockEditorEnvironment`
```suggestion
			<TestWrapper>
				<FormLayout { ...props } />
			</TestWrapper>
```

File: tests/jest/unit/modules/ai/assets/js/editor/test-utils.js [hunk 12]
```diff
@@ -32,3 +32,16 @@ export const clickEditPromptButton = () => {
...
+		global.ResizeObserver ||
+		jest.fn().mockImplementation( () => ( {
+			disconnect: jest.fn(),
+			observe: jest.fn(),
+			unobserve: jest.fn(),
+		} ) );
+};
+
+export class commonTestSetup {
+}
```
Comment (matipojo) [inline tests/jest/unit/modules/ai/assets/js/editor/test-utils.js:47]:
```suggestion
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js (hunk 11 above)
Comment (DennisNerush) [reply tests/jest/unit/modules/ai/assets/js/editor/pages/form-layout/index.test.js:63]: The mock editor is a function in the "before each" so perhaps better to leave it there. 
I currently extracted it to a method and perhaps next time we will need all 3 providers we'll extract to a component. WDYT?

**Review by matipojo — APPROVED** [review]

**Pull request merged**